lang/funcs/collection.go

```diff
@@ -634,6 +635,144 @@ var MatchkeysFunc = function.New(&function.Spec{
+// the argument sequence takes precedence.
+var MergeFunc = function.New(&function.Spec{
+	Params: []function.Parameter{},
+	VarParam: &function.Parameter{
+		Name:             "maps",
+		Type:             cty.DynamicPseudoType,
+		AllowUnknown:     true,
+		AllowDynamicType: true,
+		AllowNull:        true,
+	},
+	Type: function.StaticReturnType(cty.DynamicPseudoType),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		outputMap := make(map[string]cty.Value)
+
+		for _, arg := range args {
+
+			if !arg.Type().IsObjectType() && !arg.Type().IsMapType() {
+				return cty.NilVal, fmt.Errorf("arguments must be maps or objects, got %#v", arg.Type())
+			}
+			for it := arg.ElementIterator(); it.Next(); {
+				k, v := it.Element()
+				outputMap[k.AsString()] = v
+			}
+		}
+		return cty.ObjectVal(outputMap), nil
+	},
+})
+
+// SliceFunc contructs a function that extracts some consecutive elements
+// from within a list.
+var SliceFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.DynamicPseudoType),
+		},
+		{
+			Name: "startIndex",
+			Type: cty.Number,
+		},
+		{
+			Name: "endIndex",
+			Type: cty.Number,
+		},
+	},
+	Type: function.StaticReturnType(cty.List(cty.DynamicPseudoType)),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		inputList := args[0]
+		var startIndex, endIndex int
+
+		if err = gocty.FromCtyValue(args[1], &startIndex); err != nil {
+			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
+		}
+		if err = gocty.FromCtyValue(args[2], &endIndex); err != nil {
+			return cty.NilVal, fmt.Errorf("invalid start index: %s", err)
+		}
+
+		if startIndex < 0 {
+			return cty.NilVal, fmt.Errorf("from index must be >= 0")
+		}
+		if endIndex > inputList.LengthInt() {
+			return cty.NilVal, fmt.Errorf("to index must be <= length of the input list")
+		}
+		if startIndex > endIndex {
+			return cty.NilVal, fmt.Errorf("from index must be <= to index")
+		}
+
+		var outputList []cty.Value
+
+		i := 0
+		for it := inputList.ElementIterator(); it.Next(); {
+			_, v := it.Element()
+			if i >= startIndex && i < endIndex {
+				outputList = append(outputList, v)
+			}
+			i++
+		}
+
+		if len(outputList) == 0 {
+			return cty.ListValEmpty(cty.DynamicPseudoType), nil
+		}
+		return cty.ListVal(outputList), nil
+	},
+})
+
+// TransposeFunc contructs a function that takes a map of lists of strings and
+// swaps the keys and values to produce a new map of lists of strings.
+var TransposeFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "values",
+			Type: cty.Map(cty.List(cty.String)),
+		},
+	},
+	Type: function.StaticReturnType(cty.Map(cty.List(cty.String))),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		inputMap := args[0]
+		outputMap := make(map[string]cty.Value)
+		tmpMap := make(map[string][]string)
+
```

Comment:
The lists and strings in `inputMap` might be unknown here... but I think if we _were_ to find unknowns during the loop the best we'd be able to do is return an unknown map as the result, so might as well just do that up here before we start looping:

```go
    if !inputMap.IsWhollyKnown() {
        return cty.UnknownVal(retType), nil
    }
```